Spec edits for @defer/@stream (#742)
graphql/graphql-spec: spec/Section 6 -- Execution.md

"Executing Operations” section below.

ExecuteRequest(schema, document, operationName, variableValues, initialValue):
  Note: the execution assumes implementing language supports coroutines.
  Alternatively, the socket can provide a write buffer pointer to allow {ExecuteRequest()}
  to directly write payloads into the buffer. 
  * Let {operation} be the result of {GetOperation(document, operationName)}.
  * Let {coercedVariableValues} be the result of {CoerceVariableValues(schema, operation, variableValues)}.
  * If {operation} is a query operation:

>     * Let {payloads} be the iterator obtained from calling {ExecuteQuery(operation, schema, coercedVariableValues, initialValue, subsequentPayloads)}.
>     * For each {payload} in {payloads}:
>       * Yield {payload}.

Reviewer: This isn't guaranteed to be an iterator (`ExecuteQuery` may return an unordered map):

```suggestion
    * Let {executionResult} be the result of calling {ExecuteQuery(operation, schema, coercedVariableValues, initialValue, subsequentPayloads)}.
    * If {executionResult} is an iterator:
      * For each {payload} in {executionResult}:
        * Yield {payload}.
    * Otherwise:
      * Return {executionResult}.
```

I'm also not sure if we need an explicit return once the `For each` exits.

Further, with coroutines, are we explicitly sending the termination signal to the iterator? E.g. if the iterator is sending 1,000,000 results, and the client closes the connection after 20,000, do we need to explicitly state that this should terminate the iterator?

Author: I added details about handling termination signals in https://github.com/graphql/graphql-spec/pull/742/commits/aa2741dcca1253e4741c4924a36399b6ba10139a

"Executing Operations” section below.

ExecuteRequest(schema, document, operationName, variableValues, initialValue):

> Note: the execution assumes implementing language supports coroutines.

Reviewer 2: ```suggestion

> Note: the execution assumes implementing language supports coroutines.
```

Alternatively, the socket can provide a write buffer pointer to allow
{ExecuteRequest()} to directly write payloads into the buffer.

- Let {operation} be the result of {GetOperation(document, operationName)}.
- Let {coercedVariableValues} be the result of {CoerceVariableValues(schema,
  operation, variableValues)}.
- If {operation} is a query operation:

>   - Let {executionResult} be the result of calling {ExecuteQuery(operation,

Reviewer 3: @robrichard Is it possible this edit is not necessary considering the changes to ExecuteQuery below? Or, is there a different reason why queries are treated differently than mutations in this section?

ExecuteQuery(query, schema, variableValues, initialValue):

  * Let {subsequentPayloads} be an empty list.
  * Let {queryType} be the root Query type in {schema}.
  * Assert: {queryType} is an Object type.
  * Let {selectionSet} be the top level Selection Set in {query}.
  * Let {data} be the result of running
    {ExecuteSelectionSet(selectionSet, queryType, initialValue, variableValues, subsequentPayloads)}
    *normally* (allowing parallelization).
  * Let {errors} be any *field errors* produced while executing the
    selection set.
  * If {subsequentPayloads} is empty:
    * Return an unordered map containing {data} and {errors}.
  * If {subsequentPayloads} is not empty:
    * Yield an unordered map containing {data}, {errors}, and an entry named {hasNext} with the value {true}.

>     * Let {iterator} be the result of running YieldSubsequentPayloads(subsequentPayloads).

Reviewer: ```suggestion
    * Let {iterator} be the result of running {YieldSubsequentPayloads(subsequentPayloads)}.
```

ExecuteQuery(query, schema, variableValues, initialValue):

  * Let {subsequentPayloads} be an empty list.
  * Let {queryType} be the root Query type in {schema}.
  * Assert: {queryType} is an Object type.
  * Let {selectionSet} be the top level Selection Set in {query}.
  * Let {data} be the result of running
    {ExecuteSelectionSet(selectionSet, queryType, initialValue, variableValues, subsequentPayloads)}
    *normally* (allowing parallelization).
  * Let {errors} be any *field errors* produced while executing the
    selection set.

>   * Yield an unordered map containing {data} and {errors}.

Reviewer 4: and `hasNext: true` if `subsequentPayloads` is non-empty

  * For each {payload} in {subsequentPayloads}:
    * If {payload} is a Deferred Fragment Record:
      * Yield the value from calling {ResolveDeferredFragmentRecord(payload, variableValues, subsequentPayloads)}.
      * If {payload} is not the final payload in {subsequentPayloads}
        * Add an entry to {payload} named `hasNext` with the value {true}.
      * If {payload} is the final payload in {subsequentPayloads}
        * Add an entry to {payload} named `hasNext` with the value {false}.
    * If {payload} is a Stream Record:

>       * Yield ResolveStreamRecord(payload, variableValues, subsequentPayloads).

Reviewer 4: this needs similar logic for when to include `hasNext` as the defer case

Author: Updated this so it should be clear that the logic applies to both the `defer` and `stream` case.

    selection set.
  * Yield an unordered map containing {data}, {errors}, and {hasNext} if 
    {subsequentPayloads} is non-empty.
  * While {subsequentPayloads} is not empty:
    * Let {record} be the first complete item in {subsequentPayloads}.
    * Remove {record} from {subsequentPayloads}.
    * If {record} is a Deferred Fragment Record:
      * Let {payload} be the result of running {ResolveDeferredFragmentRecord(record, variableValues, subsequentPayloads)}.
    * If {record} is a Stream Record:
      * Let {payload} be the result of running ResolveStreamRecord(record, variableValues, subsequentPayloads).
    * If {payload} is {null}:
      * If {subsequentPayloads} is empty:
        * Yield a map containing a field `hasNext` with the value {false};
      * If {subsequentPayloads} is not empty:
        * Continue to the next record in {subsequentPayloads}.
    * If {record} is not the final element in {subsequentPayloads}
      * Add an entry to {payload} named `hasNext` with the value {true}.
    * If {record} is the final element in {subsequentPayloads}
      * Add an entry to {payload} named `hasNext` with the value {false}.

>     * Yield {payload}

Author: This case where we yield a payload of only `{ hasNext: false }` can only occur when an async iterable (or equivalent) stream record is the last value in subsequent payloads. I tried to cover it by adding this to `ResolveStreamRecord`:

  * Wait for the next item from {iterator}.
  * If an item is not retrieved because {iterator} has completed: // (I just updated this line to clarify it)
    * Return {null}
  * Let {item} be the item retrieved from {iterator}.
  * // rest of logic
  * Append {streamRecord} to {subsequentPayloads}.
  * Return {payload}.

So unless we know the underlying iterable from the stream has completed, it will be re-added to `{subsequentPayloads}` before it's processed again.

Then, the `{null}` case is handled above this section:

* If {payload} is {null}:
      * If {subsequentPayloads} is empty:
        * Yield a map containing a field `hasNext` with the value {false};
      * If {subsequentPayloads} is not empty:
        * Continue to the next record in {subsequentPayloads}.

>     * Yield {payload}

Author: I did miss adding this to the mutation and subscription section, which I will update

>     * Yield {payload}

Author: I extracted this out to a function `YieldSubsequentPayloads` which is referenced in `ExecuteQuery`, `ExecuteMutation` and `ExecuteSubscriptionEvent`

ExecuteMutation(mutation, schema, variableValues, initialValue):

  * Let {subsequentPayloads} be an empty list.
  * Let {mutationType} be the root Mutation type in {schema}.
  * Assert: {mutationType} is an Object type.
  * Let {selectionSet} be the top level Selection Set in {mutation}.
  * Let {data} be the result of running
    {ExecuteSelectionSet(selectionSet, mutationType, initialValue, variableValues, subsequentPayloads)}
    *serially*.
  * Let {errors} be any *field errors* produced while executing the
    selection set.
  * Yield an unordered map containing {data}, {errors}, and {hasNext} if 
    {subsequentPayloads} is non-empty.
  * For each {record} in {subsequentPayloads}:
    * If {record} is a Deferred Fragment Record:
      * Let {payload} be the result of running {ResolveDeferredFragmentRecord(record, variableValues, subsequentPayloads)}.
    * If {record} is a Stream Record:

>       * Let {payload} be the result of running ResolveStreamRecord(record, variableValues, subsequentPayloads).

Reviewer: (Same mod needed in more places)

>     * If {record} is not the final element in {subsequentPayloads}

Reviewer: Same comment as above regarding not being able to determine this.

Reviewer: (Same comment in more places)

  * Cancel {responseStream}

## Yield Subsequent Payloads

If an operation contains subsequent payload records resulting from `@stream` or
`@defer` directives, the {YieldSubsequentPayloads} algorithm defines how the
payloads should be processed.

YieldSubsequentPayloads(subsequentPayloads):

  * While {subsequentPayloads} is not empty:
  * Let {record} be the first complete item in {subsequentPayloads}.
    * Remove {record} from {subsequentPayloads}.
    * If {record} is a Deferred Fragment Record:
      * Let {payload} be the result of running {ResolveDeferredFragmentRecord(record, variableValues, subsequentPayloads)}.
    * If {record} is a Stream Record:
      * Let {payload} be the result of running {ResolveStreamRecord(record, variableValues, subsequentPayloads)}.

>     * If {payload} is {null}:

Reviewer: `payload` may be undefined here, if record is neither a stream or deferred fragment record?

Author: Each `record` in `subsequentPayloads` has to be either a stream record or a deferred fragment record.

Reviewer: Currently, yeah. Should we put an assert here to enforce that so that future spec edits don't accidentally break this assumption?

      * If {subsequentPayloads} is empty:

>         * Yield a map containing a field `hasNext` with the value {false};

Reviewer: And stop here, right?

```suggestion
        * Yield a map containing a field `hasNext` with the value {false}.
        * Return.
```

- Cancel {responseStream}

## Yield Subsequent Payloads

If an operation contains subsequent payload records resulting from `@stream` or
`@defer` directives, the {YieldSubsequentPayloads} algorithm defines how the
payloads should be processed.

YieldSubsequentPayloads(initialResponse, subsequentPayloads):

- Let {initialRecords} be any items in {subsequentPayloads} with a completed
  {dataExecution}.
- Initialize {initialIncremental} to an empty list.
- For each {record} in {initialRecords}:
  - Remove {record} from {subsequentPayloads}.
  - If {isCompletedIterator} on {record} is {true}:
    - Continue to the next record in {records}.
  - Let {payload} be the completed result returned by {dataExecution}.
  - Append {payload} to {initialIncremental}.
- If {initialIncremental} is not empty:
  - Add an entry to {initialResponse} named `incremental` containing the value
    {incremental}.
- Yield {initialResponse}.

> - While {subsequentPayloads} is not empty:

Reviewer 5: Everything under this needs to be indented more?

Author: updated

- If a termination signal is received:
  - For each {record} in {subsequentPayloads}:
    - If {record} contains {iterator}:
      - Send a termination signal to {iterator}.
  - Return.
- Wait for at least one record in {subsequentPayloads} to have a completed
  {dataExecution}.
- Let {subsequentResponse} be an unordered map with an entry {incremental}
  initialized to an empty list.
- Let {records} be the items in {subsequentPayloads} with a completed
  {dataExecution}.
  - For each {record} in {records}:
    - Remove {record} from {subsequentPayloads}.
    - If {isCompletedIterator} on {record} is {true}:
      - Continue to the next record in {records}.
    - Let {payload} be the completed result returned by {dataExecution}.

>     - Append {payload} to {incremental}.

Reviewer 5: This should say the `{incremental}` entry on `{subsequentReponse}` or something?

Author: updated

  - If {subsequentPayloads} is empty:

>     - Add an entry to {response} named `hasNext` with the value {false}.

Reviewer 5: ```suggestion
    - Add an entry to {subsequentResponse} named `hasNext` with the value {false}.
```

>   - Yield {response}

Reviewer 5: ```suggestion
  - Yield {subsequentResponse}
```

Author: updated

If this occurs, any sibling fields which have not yet executed or have not yet
yielded a value may be cancelled to avoid unnecessary work.

Additionally, async payload records in {subsequentPayloads} must be filtered if
their path points to a location that has resolved to {null} due to propagation
of a field error. This is described in
[Filter Subsequent Payloads](#sec-Filter-Subsequent-Payloads). These async
payload records must be removed from {subsequentPayloads} and their result must
not be sent to clients. If these async records have not yet executed or have not
yet yielded a value they may also be cancelled to avoid unnecessary work.

Note: See [Handling Field Errors](#sec-Handling-Field-Errors) for more about
this behavior.

### Filter Subsequent Payloads

When a field error is raised, there may be async payload records in
{subsequentPayloads} with a path that points to a location that has been removed
or set to null due to null propagation. These async payload records must be
removed from subsequent payloads and their results must not be sent to clients.

> In {FilterSubsequentPayloads}, {nullPath} is the path which has resolved to null

Reviewer 6: what would be the expected behavior if the asyncRecord is under a null inside a list?
With a query like

```graphql
{
  products {
    id
     ... @defer {
       name
     }
   }
 }
 ```
 
 If we get as first payload:
 
 ```json
{
  "products": [
    null,
    {
      "id": 1
    }
  ]
 }
 ```
 
 we would expect the second payload to be:
 
 ```json
 {
  "incremental":[
    {
      "path": [
        "products",
        0
      ],
      "data": null
    },
    {
      "path": [
        "products",
        1
      ],
      "data": {
        "name": "chair"
      }
    }
  ]
 }
 ```
 
 But if the payload was:
 
 ```json
{
  "products": [
    null,
    null
  ]
 }
 ```
 
 Would we cancel the second payload or return:
 
 ```json
 {
  "incremental":[
    {
      "path": [
        "products",
        0
      ],
      "data": null
    },
    {
      "path": [
        "products",
        1
      ],
      "data": null
    }
  ]
 }
```

_field error_ then that error must propagate to this entire selection set,
either resolving to {null} if allowed or further propagated to a parent field.

> If this occurs, any defer or stream executions with a path that starts with the

Reviewer 5: This is really "starts with but is not equal to" or something, right? A literal reading here would cause one to assume that the actual defer itself (whose path starts with itself) should not be returned to the client.

Also "same path as the resolved null" — this is a bit confusing as to whether it means the original null or a bubbled-up null.

Maybe both issues can be fixed while being clear that the paths you're comparing are the root paths of each defer/stream?

Author: I updated this again. Hopefully it's clearer now.

Author: I also added examples for all of the non-null cases

Author: I have now also added the complete algorithm for FilterSubsequentPayloads as implemented in https://github.com/graphql/graphql-js/pull/3720

        with the next {selection} in {selectionSet}.
      * Let {fragmentSelectionSet} be the top-level selection set of {selection}.
      * If {InlineFragment} provides the directive `@defer`, let {deferDirective} be that directive.
        * If {deferDirective}'s {if} argument is {true} or is a variable in {variableValues} with the value {true}:
          * Let {deferredFragment} be the result of calling {DeferFragment(objectType, objectValue, fragmentSelectionSet, parentPath)}.
          * Append {deferredFragment} to {deferredFragments}.
          * Continue with the next {selection} in {selectionSet}.
      * Let {fragmentGroupedFieldSet} be the result of calling {CollectFields(objectType, fragmentSelectionSet, variableValues, visitedFragments, parentPath)}.
      * For each {fragmentGroup} in {fragmentGroupedFieldSet}:
        * Let {responseKey} be the response key shared by all fields in {fragmentGroup}.
        * Let {groupForResponseKey} be the list in {groupedFields} for
          {responseKey}; if no such list exists, create it as an empty list.
        * Append all items in {fragmentGroup} to {groupForResponseKey}.
  * Return {groupedFields}.

>   

Reviewer 2: Also here, there are two spaces which lead to the note being pulled up on build:

<img width="235" alt="image" src="https://user-images.githubusercontent.com/9714350/93991630-cde7f100-fd8c-11ea-81d7-2c6ba15e4d13.png">

        with the next {selection} in {selectionSet}.
      * Let {fragmentSelectionSet} be the top-level selection set of {fragment}.
      * If {fragmentSpreadName} provides the directive `@defer`, let {deferDirective} be that directive.
        * If {deferDirective}'s {if} argument is {true} or is a variable in {variableValues} with the value {true}:
          * Let {deferredFragment} be the result of calling {DeferFragment(objectType, objectValue, fragmentSelectionSet, parentPath)}.
          * Append {deferredFragment} to {deferredFragments}.
          * Continue with the next {selection} in {selectionSet}.
      * Let {fragmentGroupedFieldSet} be the result of calling
        {CollectFields(objectType, fragmentSelectionSet, variableValues, visitedFragments)}.
      * For each {fragmentGroup} in {fragmentGroupedFieldSet}:
        * Let {responseKey} be the response key shared by all fields in {fragmentGroup}.
        * Let {groupForResponseKey} be the list in {groupedFields} for
          {responseKey}; if no such list exists, create it as an empty list.
        * Append all items in {fragmentGroup} to {groupForResponseKey}.

>       * Add {fragmentSpreadName} to {visitedFragments}.

Reviewer: Doesn't this change open the possibility of infinite recursion because you visit the fragment before you mark that it was visited?

Author: i updated this a bit and now only fragments that have `@defer` will skip the check for visited fragments, which does leave open the possibility of infinite recursion, but I think that should be caught by validation rules? Another option is to make this into something that keys on both the fragment name and the defer directive.

    * if {objectType} is an implementation of {fragmentType}, return {true} otherwise return {false}.
  * If {fragmentType} is a Union:
    * if {objectType} is a possible type of {fragmentType}, return {true} otherwise return {false}.

>  

Reviewer 2: There is a space in this line that messes up building the spec.

When the spec is being build the space leads to the indentation of DeferFragment.

Reviewer 2: <img width="185" alt="image" src="https://user-images.githubusercontent.com/9714350/93991304-69c52d00-fd8c-11ea-8537-a6cbbb5cba42.png">

    - Append {selection} to the {groupForResponseKey}.
  - If {selection} is a {FragmentSpread}:
    - Let {fragmentSpreadName} be the name of {selection}.

>     - If {fragmentSpreadName} provides the directive `@defer` and it's {if}

Reviewer 7: Typo
```suggestion
    - If {fragmentSpreadName} provides the directive `@defer` and its {if}
```

Reviewer 8: Yeah don't know about that 🤔

Author: Thanks!

  * If {fragmentType} is a Union:
    * if {objectType} is a possible type of {fragmentType}, return {true} otherwise return {false}.

DeferFragment(objectType, objectValue, fragmentSelectionSet, parentPath):
  * Let {label} be the value or the variable to {deferDirective}'s {label} argument.
  * Let {deferredFragmentRecord} be the result of calling {CreateDeferredFragmentRecord(label, objectType, objectValue, fragmentSelectionSet, parentPath)}.
  * return {deferredFragmentRecord}.

#### Deferred Fragment Record

**Formal Specification**
Let {deferredFragmentRecord} be an inline fragment or fragment spread with `@defer` provided.
Deferred Fragment Record is a structure containing: 
* {label} value derived from the `@defer` directive.
* {objectType} of the {deferredFragmentRecord}.
* {objectValue} of the {deferredFragmentRecord}.
* {fragmentSelectionSet}: the top level selection set of {deferredFragmentRecord}.

> * {path} a list of field names and indices from root to {deferredFragmentRecord}.

Reviewer: Formatting does not match rest of document.

    - Append {selection} to the {groupForResponseKey}.
  - If {selection} is a {FragmentSpread}:
    - Let {fragmentSpreadName} be the name of {selection}.
    - If {fragmentSpreadName} provides the directive `@defer` and its {if}
      argument is {true} or is a variable in {variableValues} with the value
      {true}:

>       - Let {deferDirective} be that directive.

Reviewer: This currently implies that `... @defer { __typename }` will result in the selection set _NOT_ being deferred since it does not have a `true` `if` argument (and this argument has no default).

Author: I rewrote this, inverting the logic, which is closer to what is used for for the include directive further up in this function.

>  If {fragmentSpreadName} provides the directive `@defer` and its {if}
      argument is not {false} or is a variable in {variableValues} with the
      value {false}:

That should make it clear that only `if: false` will prevent the fragment from being deferred

    - Append {selection} to the {groupForResponseKey}.
  - If {selection} is a {FragmentSpread}:
    - Let {fragmentSpreadName} be the name of {selection}.
    - If {fragmentSpreadName} provides the directive `@defer` and its {if}

>       argument is not {false} or is a variable in {variableValues} with the

Reviewer 5: ```suggestion
      argument is not {false} or a variable in {variableValues} with the
```

I think this reads better — otherwise it feels like it's saying "(is not false) or (is a variable with the value false)"

Author: I updated to use the same wording used for `includeDirective`

      fragmentType)} is false, continue with the next {selection} in
      {selectionSet}.
    - Let {fragmentSelectionSet} be the top-level selection set of {selection}.
    - If {InlineFragment} provides the directive `@defer` and its {if} argument

>       is not {false} or is a variable in {variableValues} with the value

Reviewer 5: ```suggestion
      is not {false} or a variable in {variableValues} with the value
```

  {path}: a list of field names and indices from root to {streamField}.
  {fields}: the group of fields grouped by CollectFields() for {streamField}.
  {innerType}: inner type of {streamField}'s type.

CreateStreamRecord(label, initialCount, iterator, resolvedItems, index, fields, innerType):
*  Construct a stream record based on the parameters passed in.

ResolveStreamRecord(streamRecord, variableValues, subsequentPayloads):
  * Let {label}, {iterator}, {resolvedItems}, {index}, {path}, {fields},
    {innerType} be the correspondent fields on the Stream Record structure.
  * Remove the first entry from {resolvedItem}, let the entry be {item}. If 
    {resolvedItem} is empty, retrieve more items from {iterator}:
    * Append {index} to {path}.
    * Increment {index}.
    * Let {payload} be the result of calling CompleteValue(innerType, fields, item, variableValues, subsequentPayloads, path)}.
    * Add an entry to {payload} named `label` with the value {label}.
    * Add an entry to {payload} named `path` with the value {path}.
    * If {resolveItem} is not empty or {iterator} does not reach the end:
      * Append {streamRecord} to {subsequentPayloads}.

>     * Return {payload}.

Reviewer: I think you might need to revisit this definition; I might be reading it wrong but it feels like you may have used `resolvedItem` in the place of `resolvedItems` in one place, and the "retrieve more items from {iterator}" isn't super clear, it refers to "If {resolvedItem} is not empty" under an item which already seems to state "If {resolvedItem} is empty", and there doesn't seem to be a point at which `{resolvedItem}` would change between the two. Also it uses `resolveItem` rather than `resolvedItem` which I think is a typo since it doesn't seem to be referred to elsewhere.

While nearly all of GraphQL execution can be described generically, ultimately
the internal system exposing the GraphQL interface must provide values.
This is exposed via {ResolveFieldValue}, which produces a value for a given
field on a type for a real value. In addition, {ResolveFieldGenerator} will be 
exposed to produce an iterator for a field with `List` return type.
The internal system may optionally define a generator function. In the case
where the generator is not defined, the GraphQL executor provides a default generator.
For example, a trivial generator that yields the entire list upon the first iteration.

As an example, a {ResolveFieldValue} might accept the {objectType} `Person`, the {field}
{"soulMate"}, and the {objectValue} representing John Lennon. It would be
expected to yield the value representing Yoko Ono.

A {ResolveFieldGenerator} might accept the {objectType} `MusicBand`, the {field} 
{"members"}, and the {objectValue} representing Beatles. It would be expected to yield

> a iterator of values representing, John Lennon, Paul McCartney, Ringo Starr and 

Reviewer 9: type: the comma after `representing`

George Harrison.

ResolveFieldValue(objectType, objectValue, fieldName, argumentValues):
  * Let {resolver} be the internal function provided by {objectType} for
    determining the resolved value of a field named {fieldName}.
  * Return the result of calling {resolver}, providing {objectValue} and {argumentValues}.
  
ResolveFieldGenerator(objectType, objectValue, fieldName, argumentValues, initialCount):
  * If {objectType} provide an internal function {generatorResolver} for
    generating partitially resolved value of a list field named {fieldName}:
    * Let {generatorResolver} be the internal function.
    * Return the iterator from calling {generatorResolver}, providing
      {objectValue}, {argumentValues} and {initialCount}.
  * Create {generator} from {ResolveFieldValue(objectType, objectValue, fieldName, argumentValues)}.
  * Return {generator}.

Note: It is common for {resolver} to be asynchronous due to relying on reading
an underlying database or networked service to produce a value. This
necessitates the rest of a GraphQL executor to handle an asynchronous

> execution flow. In addition, a commom implementation of {generator} is to leverage 

Reviewer 9: typo `commom`

asynchronous iterators or asynchronous generators provided by many programming languages.

### Value Completion

After resolving the value for a field, it is completed by ensuring it adheres
to the expected return type. If the return type is another Object type, then
the field execution process continues recursively. In the case where a value
returned for a list type field is an iterator due to `@stream` specified on the 

> field, value completition iterates over the iterator until the number of items

Reviewer 9: type `completition`

  * If {result} is {null} (or another internal value similar to {null} such as
    {undefined} or {NaN}), return {null}.
  * If {fieldType} is a List type:
    * If {result} is an iterator:
      * Let {field} be the first entry in {fields}.
      * Let {innerType} be the inner type of {fieldType}.
      * Let {streamDirective} be the `@stream` directived provided on {field}.
      * Let {initialCount} be the value or variable provided to {streamDirective}'s {initialCount} argument.
      * Let {label} be the value or variable provided to {streamDirective}'s {label} argument.
      * Let {resolvedItems} be an empty list
      * For each {members} in {result}:
        * Append all items from {members} to {resolvedItems}.
        * If the length of {resolvedItems} is greater or equal to {initialCount}:
          * Let {initialItems} be the sublist from the beginning to the {initialCount}-th item.

>           * Let {remainingItems} be the sublist from {initialCount}-th + 1 item to the end.

Reviewer: Do you mean this?
```suggestion
          * Let {remainingItems} be the sublist from {initialCount + 1}-th item to the end.
```

>           * Let {initialItems} be the sublist from the beginning to the {initialCount}-th item.
>           * Let {remainingItems} be the sublist from {initialCount}-th + 1 item to the end.

Reviewer: Might be clearer to go with something that doesn't depend on indexes (which could be 0- or 1- indexed), such as:

```suggestion
          * Let {initialItems} be the sublist of the first {initialCount} items from {resolvedItems}.
          * Let {remainingItems} be the sublist of the items in {resolvedItems} after the first {initialCount} items.
```

  * If {result} is {null} (or another internal value similar to {null} such as
    {undefined} or {NaN}), return {null}.
  * If {fieldType} is a List type:

>     * If {result} is an iterator:

Reviewer 2: We have now a rough implementation of stream for hotchocolate and I ran into some things that are not so clear. 

When streaming a field that has a list of composite types the iterator will yield objects like the following:

```json
{
  "id": "1000",
   "name": "Luke Skywalker"
}
```

These will than be integrated into a result that could look like the following:

```json
{
  "path": ["hero", "friends", "nodes", 1],
  "data": {
    "id": "1000",
    "name": "Luke Skywalker"
  },
  "hasNext": false
}
```

This works quite well... But what if the list is a list of leaf values like a list of strings? I have the same issue with a nested list where the element is actually a list of something.

Can data in this case be a scalar or a list? Would this break the contract for GraphQL results?

Reviewer 8: A list of strings would just be:

```json
{
  "path": ["hero", "friends", "items"],
  "data": ["hi there"],
  "hasNext": false
}
```

i suspect youre talking about `@stream` there — then:

```json
{
  "path": ["hero", "friends", "nodes", 0],
  "data": "hi there",
  "hasNext": false
}
```

Reviewer 2: sure... but so far `data` was guaranteed by the spec to be `null` or a map. We need to clarify this.

Reviewer 2: Maybe we should also put some examples in for people who are implementing this.

Author: I opened a discussion for this here: https://github.com/robrichard/defer-stream-wg/discussions/21

    - Add an entry to {payload} named `errors` with the value {errors}.
  - Add an entry to {payload} named `items` with a list containing the value
    {data}.
  - Add an entry to {payload} named `label` with the value {label}.
  - Add an entry to {payload} named `path` with the value {itemPath}.
  - Return {payload}.
- Set {dataExecution} on {streamRecord}.
- Append {streamRecord} to {subsequentPayloads}.

CompleteValue(fieldType, fields, result, variableValues, path,
subsequentPayloads, asyncRecord):

- If the {fieldType} is a Non-Null type:
  - Let {innerType} be the inner type of {fieldType}.
  - Let {completedResult} be the result of calling {CompleteValue(innerType,
    fields, result, variableValues, path)}.
  - If {completedResult} is {null}, raise a _field error_.
  - Return {completedResult}.
- If {result} is {null} (or another internal value similar to {null} such as
  {undefined}), return {null}.
- If {fieldType} is a List type:
  - If {result} is an iterator:
    - Let {field} be the first entry in {fields}.
    - Let {innerType} be the inner type of {fieldType}.
    - If {field} provides the directive `@stream` and its {if} argument is not

>       {false} or is a variable in {variableValues} with the value {false} and

Reviewer 5: ```suggestion
      {false} or a variable in {variableValues} with the value {false} and
```
